### PR TITLE
[#2PlaysAMonth] Digits Delight

### LINKED ISSUE #990: [#2PlaysAMonth]: Digits Delight: Delight in the magic of numbers with Digits Delight's amazing facts!
### What's the Play about?

This play is going to be a fun and learning play, where we can explore facts about numbers. Here we can get facts about particular numbers, dates, and years. We can also investigate random number facts with just one click. 

### What ReactJS concept will be used to create this play?

- useState
- useEffect
- Axios

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Shivam-Katare**
# Description

Digits Delight is a fun and learning play, where we can explore facts about numbers. Here we can get facts about particular numbers, dates, and years. We can also investigate random number facts with just one click.

Fixes #990 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# Screenshots or example output

Screenshot (778)
Screenshot (779)
Screenshot (780)

## Demo

- [Video](https://www.stack-stream.com/v/introducing-digits-delight-for-2-plays-a-month)

### CODE REVIEW

File: src/plays/digital-delight/DigitsDelight.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+// WARNING: Do not change the entry componenet name
+function DigitsDelight(props) {
+  // Your Code Start below.
+
+  const [showModal, setShowModal] = useState(false);
+
+  const toggle = () => {
+    setShowModal(!showModal);
+  };
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/DigitsDelight.js:18]:
```suggestion
  const toggle = () => setShowModal(prev => !prev);
  
```

File: src/plays/digital-delight/Readme.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Number Facts
+
+Get interesting facts about numbers.
+
+## Play Demographic
+
+- Language: js
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/Readme.md:7]:
```suggestion
- Language: JavaScript
```

File: src/plays/digital-delight/components/CategoryFact.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import React, { useState } from 'react';
+import axios from 'axios';
+import Math from '../components/images/math.png';
+import Year from '../components/images/year.png';
+import Date from '../components/images/date.png';
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:5]:
```suggestion
import MathImage from '../components/images/math.png';
import YearImage from '../components/images/year.png';
import DateImage from '../components/images/date.png';
```

File: src/plays/digital-delight/components/CategoryFact.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import './Categoryfact.css';
+
+function CategoryFact() {
+  const [fact, setFact] = useState('Click on buttons below to get facts by category');
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:9]: Move this to constant
```
'Click on buttons below to get facts by category'
```

```suggestion
  const [fact, setFact] = useState('Click on buttons below to get facts by category');
```

File: src/plays/digital-delight/components/CategoryFact.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  const speechHandler = (msg) => {
+    // This for loop helps to give the selected voice
+    for (let i = 0; i < window.speechSynthesis.getVoices().length; i++) {
+      if (window.speechSynthesis.getVoices()[i].name === selectedVoice) {
+        msg.voice = window.speechSynthesis.getVoices()[i];
+      }
+    }
+    window.speechSynthesis.cancel();
+    window.speechSynthesis.speak(msg);
+  };
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:21]:
```suggestion
  const speechHandler = (msg) => {
    let voices = [];
window.speechSynthesis.onvoiceschanged = () => {
  voices = window.speechSynthesis.getVoices();
};
    // This for loop helps to give the selected voice
    for (let i = 0; i < voices.length; i++) {
      if (voices[i].name === selectedVoice) {
        msg.voice = voices[i];
      }
    }
    window.speechSynthesis.cancel();
    window.speechSynthesis.speak(msg);
  };
```

File: src/plays/digital-delight/components/CategoryFact.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      headers: {
+        'X-RapidAPI-Key': process.env.REACT_APP_DIGITSDELIGHT_APIKEY,
+        'X-RapidAPI-Host': 'numbersapi.p.rapidapi.com'
+      }
+    };
+    try {
+      const response = await axios.request(options);
+      setFact(response.data.number + ' is ' + response.data.text);
+      speechHandler(
+        new SpeechSynthesisUtterance(response.data.number + ' is ' + response.data.text)
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:37]:
```suggestion
      const { data } = await axios.request(options);
      setFact(data.number + ' is ' + data.text);
      speechHandler(
        new SpeechSynthesisUtterance(data.number + ' is ' + data.text)
```

File: src/plays/digital-delight/components/CategoryFact.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      );
+      setShowSelecter(true);
+    } catch (error) {
+      console.error(error);
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:41]: @Shivam-Katare , I would recommend not including console log statements.

File: src/plays/digital-delight/components/CategoryFact.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    }
+  };
+
+  const handleVoiceChange = (event) => {
+    setSelectedVoice(event.target.value);
+  };
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:47]:
```suggestion
  const handleVoiceChange = ({target}) => setSelectedVoice(target.value);
 
```

File: src/plays/digital-delight/components/CategoryFact.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+  return (
+    <>
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:50]: If not needed then remove this fragment.

File: src/plays/digital-delight/components/CategoryFact.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      <div className="digits-delight-voice-selection-container" style={{ marginBottom: '-25px' }}>
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/CategoryFact.js:51]: @Shivam-Katare, I would not recommend the negative margins, see if another way is possible.

File: src/plays/digital-delight/components/Modal.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+      info: 'The 1st one is the random fact section, where, on clicking on the random button, you will get 3 random facts about numbers.'
+    },
+    {
+      info: 'The 2nd one is Number Fact Section, where you will enter the number, and on clicking on the `get facts` button, you will get the fact about that number.'
+    },
+    {
+      info: 'In the 3rd and the last category, you will get on scrolling a bit. Here, you can listen to facts by different categories and you can listen to them by choosing different voices.'
+    },
+    {
+      info: 'There are infinite numbers present in the world, so there can be a possibility to not have a fact about it so for that, this app will roundd off that number and tells you the fact about that number.'
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/Modal.js:27]: Move this to constant file.

File: src/plays/digital-delight/components/Modal.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+  useEffect(() => {
+    setCurrentState(0);
+  }, [showModal]);
+
+  const buttonHandler = (val) => () => {
+    if ((currState === 0 && val < 0) || (currState === 6 && val > 0)) return;
+    setCurrentState(currState + val);
+  };
+  if (!showModal) return false;
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/Modal.js:42]:
```suggestion
  if (!showModal) return null;
```

File: src/plays/digital-delight/components/Random.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
+import React, { useState } from 'react';
+import axios from 'axios';
+import './Random.css';
+
+function Random() {
+  const [isFactsLoading, setIsFactsLoading] = useState(false);
+  const [facts, setFacts] = useState([
+    'Discover fascinating number facts with just one click!',
+    'Enter any number and uncover fun and educational facts about it.',
+    'Keep scrolling to find even more intriguing number trivia to amaze your friends with!'
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/Random.js:10]: Move this to a constant file.

File: src/plays/digital-delight/components/Random.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      url: 'https://numbersapi.p.rapidapi.com/random/trivia',
+      params: { min: '1', max: '999', fragment: 'true', json: 'true' },
+      headers: {
+        'X-RapidAPI-Key': process.env.REACT_APP_DIGITSDELIGHT_APIKEY,
+        'X-RapidAPI-Host': 'numbersapi.p.rapidapi.com'
+      }
+    };
+    for (let i = 0; i < 3; i++) {
+      const response = await axios.request(options);
+      fetchedFacts.push(response.data.number + ' is ' + response.data.text);
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/Random.js:27]:
```suggestion
      const { data } = await axios.request(options);
      fetchedFacts.push(data.number + ' is ' + data.text);
```

File: src/plays/digital-delight/components/Random.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    setFacts(fetchedFacts);
+    setIsFactsLoading(false);
+  }
+
+  function handleGetFactClick() {
+    fetchData();
+  }
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/Random.js:35]: @Shivam-Katare , Do we really need this? I would recommend directly passing the `fetchData` method to onClick.

File: src/plays/digital-delight/components/TextFact.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        .request(options)
+        .then((response) => {
+          setFact(response.data.number + ' is ' + response.data.text);
+          setError('');
+        })
+        .catch((error) => {
+          setError(`Error: ${error.message}`);
+          setFact('');
+        })
+        .finally(() => setLoading(false));
```
Comment (Sachin-chaurasiya) [inline src/plays/digital-delight/components/TextFact.js:33]: Can we use `async/await` here?

**Review by Sachin-chaurasiya — COMMENTED** [review]
@Shivam-Katare , Left a few comments, thanks

**Review by koustov — DISMISSED** [review]
lgtm

**Comment by Shivam-Katare** [comment]
@Sachin-chaurasiya, All issues resolved. Thanks.

**Review by Sachin-chaurasiya — APPROVED** [review]
Great work @Shivam-Katare  🙌🏻,
Thanks for the contributions ✨

**Review by koustov — APPROVED** [review]
lgtm

**Pull request merged**